### DISCUSSION TITLE
Output with ESP32-LyratT v4.3 requires PIN_FUNC_SELECT and  WRITE_PERI_REG code

### DISCUSSION THREAD

**Discussion by steindevices** [pschatzmann/arduino-audio-tools#805] [Q&A] [answered]
Hi Phil,

First, great job on the library! Makes it SO much easier to develop audio apps with the ESP32.

Anyway, I have run into an odd problem with my ESP32-LyraT where I must add the following code after the out.begin(config) to get output from my board using AudioTools:

            PIN_FUNC_SELECT(PERIPHS_IO_MUX_GPIO0_U, FUNC_GPIO0_CLK_OUT1);
            WRITE_PERI_REG(PIN_CTRL, 0xFFF0);

How I was able to track this down is that everything worked fine when testing with the AudioKit Sine generator example, but when I attempted the same thing with the AudioTools version no output. I suspected it might be a clock issue and found these two lines in the I2S AudioKit code. When I added these lines directly into the AudioTools .ino code just after out.begin(config), everything worked.

Here is the snippet from my working code:

  // start I2S
  Serial.println("starting I2S...");
  auto config = out.defaultConfig(TX_MODE);
  config.sample_rate = sample_rate; 
  config.channels = channels;
  config.bits_per_sample = 16;   
  out.begin(config);
  PIN_FUNC_SELECT(PERIPHS_IO_MUX_GPIO0_U, FUNC_GPIO0_CLK_OUT1);
  WRITE_PERI_REG(PIN_CTRL, 0xFFF0);

Any ideas?

Aaron

**Comment by pschatzmann** [comment, marked as answer]
If you use a very old ESP32 version this code is also called automatically, but this is an obsolete way to define the master clock! 
In any new version this is identical with defining the config.pin_mck = 0;
You should be able to check in the log if this is set correctly...

How can I reproduce your issue ?

**Reply by steindevices** [reply]
Good call! The problem was using an old ESP32 board version in my old Arduino environment. In case anyone else has this problem, here are the specifics to the solution.

First, I was using an old version of the Arduino IDE (1.8.12) along with an old version of the ESP-IDF code within the Arduino environment.  For my ESP32-LyraT, I needed ESP-IDF version >= 4.4 to ensure that config.pin_mck was being set correctly. The 1,8.x Arduino IDE appears to only support up to version 3.3 of the ESP-IDF environment when the ESP32 board is added using Board Manager. 

So I installed version 2.1.0 of the Arduino IDE, but this alone was not enough to allow me to update to a more recent version of the ESP32 ESP-IDF code. To do this, I had to add the URL:

_https://raw.githubusercontent.com/espressif/arduino-esp32/gh-pages/package_esp32_index.json_ 

to the Additional Boards in the Arduino IDE preferences dialog. This then allowed me to go to Board Manager in the Arduino IDE and to update my ESP32 board to the most recent ESP-IDF (5.1).

Hope this makes sense and helps out anyone else with this issue.

**Reply by pschatzmann** [reply]
The critical thing is to use the right URL! You can still use the old Arduino version which has the big advantage, that the Stack Decoder  can be used...

**Reply by steindevices** [reply]
Hi Phil,

That's interesting. I tried adding the above URL in my Arduino IDE 1.8.12 environment, but it still only allowed updating to ESP32 board version 1.0.6 which I believe is ESP-IDF 3.3. Is there another URL I should be using in that environment? The newer 2.x Arduino IDE appears to be somewhat slower and has some issues in OTA updating.

**Reply by pschatzmann** [reply]
https://espressif-docs.readthedocs-hosted.com/projects/arduino-esp32/en/latest/installing.html

**Reply by steindevices** [reply]
Thanks for the link!

I updated my preferences with the correct URL:

https://espressif.github.io/arduino-esp32/package_esp32_index.json

And now both Arduino IDEs are using the up-to-date ESP32 board (i.e. 2.0.8). I think part of the answer was removing the old ESP32 boards using Boards Manager, then updating the URL, then re-installing the ESP32 board.
